### PR TITLE
[SPARK-36879][SQL] Support Parquet v2 data page encoding (DELTA_BINARY_PACKED) for the vectorized path 

### PR DESCRIPTION
**Pull request by parthchandra**
### What changes were proposed in this pull request?
Implements a vectorized version of the parquet reader for DELTA_BINARY_PACKED encoding
This PR includes a previous PR for this issue which passed the read request thru to the parquet implementation and which was not vectorized. The current PR builds on top of that PR (hence both are included).

### Why are the changes needed?
Currently Spark throws an exception when reading data with these encodings if vectorized reader is enabled

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Additional unit tests for the encoding for both long and integer types (mirroring the unit tests in the Parquet implementation)

### CODE REVIEW

**Comment by dbtsai** [comment]
Jenkins, retest this please

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+   * mini block has a size of 8*n, unpack 8 value each time
+   * @see org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesReader#unpackMiniBlock
+   */
+  private void unpackMiniBlock() throws IOException {
+    this.unpackedValuesBuffer = new long[miniBlockSizeInValues];
+    BytePackerForLong packer = Packer.LITTLE_ENDIAN.newBytePackerForLong(
+        bitWidths[currentMiniBlock]);
+    for (int j = 0; j < miniBlockSizeInValues; j += 8) {
+      ByteBuffer buffer = in.slice(packer.getBitWidth());
+      packer.unpack8Values(buffer, buffer.position(), unpackedValuesBuffer, j);
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:302]: wonder if we can unpack 32 values at once, since the [spec](https://github.com/apache/parquet-format/blob/master/Encodings.md#delta-encoding-delta_binary_packed--5) says the number of values in a mini block is a multiple of 32.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+    //read values from miniblock
+    int valuesRead = 0;
+    for (int i = miniBlockSizeInValues - remainingInMiniBlock;
+        i < miniBlockSizeInValues && valuesRead < remaining; i++) {
+      //calculate values from deltas unpacked for current block
+      long outValue = lastValueRead + minDeltaInCurrentBlock + unpackedValuesBuffer[i];
+      lastValueRead = outValue;
+      outputWriter.write(c, rowId + valuesRead, outValue);
+      remaining--;
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:271]: hmm I wonder why we decrement `remaining` here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala [hunk 3]
```diff
@@ -122,4 +126,40 @@ class ParquetEncodingSuite extends ParquetCompatibilityTest with SharedSparkSess
...
+  test("parquet v2 pages - delta encoding") {
+    val extraOptions = Map[String, String](
+      ParquetOutputFormat.WRITER_VERSION -> ParquetProperties.WriterVersion.PARQUET_2_0.toString,
+      ParquetOutputFormat.ENABLE_DICTIONARY -> "false"
+    )
+
+    val hadoopConf = spark.sessionState.newHadoopConfWithOptions(extraOptions)
+    withSQLConf(
+      SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> "true",
+      ParquetOutputFormat.JOB_SUMMARY_LEVEL -> "ALL") {
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala:139]: why we need this?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetDeltaEncodingIntegerSuite.scala [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+import org.apache.spark.sql.execution.vectorized.{OnHeapColumnVector, WritableColumnVector}
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.IntegerType
+
+/**
+ * Read tests for vectorized Delta binary packed Integer reader.
+ * Translated from
+ *  org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesWriterForIntegerTest
+ */
+class ParquetDeltaEncodingIntegerSuite extends ParquetCompatibilityTest with SharedSparkSession {
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetDeltaEncodingIntegerSuite.scala:37]: can we combine this with `ParquetDeltaEncodingLongSuite`? we can have a base class with the common code (need to abstract a few things like how to allocate an array, how to generate a rand int/long, and Parquet's value writer class etc.)

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  @Override
+  public byte readByte() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public short readShort() {
+    return 0;
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java:51]: why return 0 here?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  @Override
+  public void readBytes(int total, WritableColumnVector c, int rowId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void readShorts(int total, WritableColumnVector c, int rowId) {
+
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java:71]: ditto

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+  public void readIntegers(int total, WritableColumnVector c, int rowId) {
+    readValues(total, c, rowId, (w, r, v) -> {
+      c.putInt(r, (int) v);
+    });
+  }
+
+  @Override
+  public void readIntegersWithRebase(int total, WritableColumnVector c, int rowId,
+      boolean failIfRebase) {
+    throw new UnsupportedOperationException("Only readIntegers is valid.");
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:128]: I think we'll need to implement these too.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+  @Override
+  public void skipShorts(int total) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void skipIntegers(int total) {
+    // Read the values but don't write them out (the writer output method is a no-op)
+    readValues(total, null, -1, (w, r, v) -> {
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:185]: I think this can be done more efficiently, for instance we don't need to unpack the bits anymore, and don't need to compute the original value from delta, etc.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  @Override
+  public void skipDoubles(int total) {
+
+  }
+
+  @Override
+  public void skipBinary(int total) {
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java:168]: we need this too

**Review by sunchao — COMMENTED** [review]
Thanks @parthchandra for working on this! I left some comments.

**Comment by sunchao** [comment]
cc @sadikovi @viirya @dongjoon-hyun too

**Comment by parthchandra** [comment]
Thank you for the review @sunchao! Let me address the comments.

File: .gitignore [hunk 10]
```diff
@@ -33,6 +33,7 @@ R/pkg/tests/fulltests/Rplots.pdf
 build/*.jar
 build/apache-maven*
 build/scala*
+build/zinc*
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply .gitignore:36]: +1 with @sunchao 's comment. Please remove this from this PR.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 11]
```diff
@@ -280,6 +277,20 @@ private void initDataReader(
     }
   }
 
+    private ValuesReader getValuesReader(Encoding encoding) {
+        switch (encoding) {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:281]: Could you use `two-space` indentation like the other part of this file, @parthchandra ?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+import org.apache.parquet.bytes.ByteBufferInputStream;
+import org.apache.parquet.bytes.BytesUtils;
+import org.apache.parquet.column.values.ValuesReader;
+import org.apache.parquet.column.values.bitpacking.BytePackerForLong;
+import org.apache.parquet.column.values.bitpacking.Packer;
+import org.apache.parquet.io.ParquetDecodingException;
+import org.apache.parquet.io.api.Binary;
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+
+import java.io.IOException;
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:30]: The import order is a little strange. Could you grouping `java` import (line 30 and 19) together as the first group?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+    double miniSize = (double) blockSizeInValues / miniBlockNumInABlock;
+    Preconditions.checkArgument(miniSize % 8 == 0,
+        "miniBlockSize must be multiple of 8, but it's " + miniSize);
+    this.miniBlockSizeInValues = (int) miniSize;
+    this.totalValueCount = BytesUtils.readUnsignedVarInt(in);
+    this.bitWidths = new int[miniBlockNumInABlock];
+
+    // read the first value
+    firstValue = BytesUtils.readZigZagVarLong(in);
+
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:80]: nit. Let's remove redundant empty line.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    // Read the values but don't write them out (the writer output method is a no-op)
+    readValues(total, null, -1, (w, r, v) -> {
+    });
+  }
+
+  @Override
+  public void skipLongs(int total) {
+    // Read the values but don't write them out (the writer output method is a no-op)
+    readValues(total, null, -1, (w, r, v) -> {
+    });
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:193]: Let's make these two lines into a single liner.
```scala
- readValues(total, null, -1, (w, r, v) -> {
- });
+ readValues(total, null, -1, (w, r, v) -> {});
```

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+    throw new UnsupportedOperationException();
+  }
+
+  private void readValues(int total, WritableColumnVector c, int rowId,
+      IntegerOutputWriter outputWriter) {
+    int remaining = total;
+    if (valuesRead + total > totalValueCount) {
+      throw new ParquetDecodingException(
+          "no more values to read, total value count is " + valuesRead);
+    }
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:222]: Could you put this at the beginning before `int remaining = total;`?

**Comment by parthchandra** [comment]
Looks like it may take some time to address some of the review comments. Marking this PR as draft in the meantime.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetTest.scala [hunk 16]
```diff
@@ -168,6 +169,16 @@ private[sql] trait ParquetTest extends FileBasedDataSourceTest {
     withSQLConf(SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> "true")(code)
   }
 
+  def withAllParquetWriters(code: => Unit): Unit = {
+    ParquetOutputFormat.WRITER_VERSION -> ParquetProperties.WriterVersion.PARQUET_2_0.toString
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetTest.scala:173]: what is this line for?

**Review by sunchao — COMMENTED** [review]
Looks much better now @parthchandra ! do you have any benchmark result to share, i.e., comparing the vectorized reader versus non-vectorized reader from `parquet-mr`? you can use `DataSourceReadBenchmark` for this.

**Comment by parthchandra** [comment]
> Looks much better now @parthchandra ! do you have any benchmark result to share, i.e., comparing the vectorized reader versus non-vectorized reader from `parquet-mr`? you can use `DataSourceReadBenchmark` for this.

I've added a couple of cases to the benchmark for this. Because Delta encoding is more suited for sorted data, I also ran a modified version (not checked in) on sorted data. 

TL;DR For unsorted data Delta vectorized is slower than default (RLE) vectorized reader. For sorted data, Delta vectorized is much faster than unsorted data, but is still slightly slower than default (RLE) vectorized reader. For both unsorted/sorted data the vectorized reader is ~7x/~10x faster than the Parquet-MR reader.

Here is the summary for INT data

UNSORTED
```
OpenJDK 64-Bit Server VM 11.0.6+10-LTS on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz

SQL Single INT Column Scan:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL CSV                                           12625          12720         135          1.2         802.7       1.0X
SQL Json                                           6784           6844          85          2.3         431.3       1.9X
SQL Parquet Vectorized                              106            112           7        148.5           6.7     119.2X
SQL Parquet MR                                     1594           1601           9          9.9         101.4       7.9X
SQL ORC Vectorized                                  218            227           5         72.0          13.9      57.8X
SQL ORC MR                                         1289           1320          44         12.2          81.9       9.8X
SQL Parquet Vectorized (Delta Binary)               232            243          17         67.9          14.7      54.5X
SQL Parquet MR (Delta Binary)                      1462           1478          24         10.8          92.9       8.6X
```

SORTED 
```
OpenJDK 64-Bit Server VM 11.0.6+10-LTS on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
SQL Single INT Column Scan:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL CSV                                           12135          12150          22          1.3         771.5       1.0X
SQL Json                                           6138           6177          54          2.6         390.3       2.0X
SQL Parquet Vectorized                              102            107           7        153.5           6.5     118.4X
SQL Parquet MR                                     1489           1496           9         10.6          94.7       8.1X
SQL ORC Vectorized                                  166            171           3         94.9          10.5      73.2X
SQL ORC MR                                         1184           1188           5         13.3          75.3      10.3X
SQL Parquet Vectorized (Delta Binary)               130            138           8        120.7           8.3      93.1X
SQL Parquet MR (Delta Binary)                      1287           1306          26         12.2          81.8       9.4X
```

**Comment by parthchandra** [comment]
I also checked on a real-world dataset. After writing the data compression was marginally better.
Parquet v1 (RLE) = 301.9 GB, Parquet v2 (DELTA_BINARY, DELTA_BYTE_ARRAY) = 299.9 GB. The bulk of the data was String data so this does not reflect the effectiveness of  DELTA_BINARY since most of the data would be encoded using DELTA_BYTE_ARRAY. 
Haven't been able to run a read benchmark on this yet.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+    readValues(1, null, 0, (w, r, v) -> intVal = (int) v);
+    return intVal;
+  }
+
+  @Override
+  public long readLong() {
+    readValues(1, null, 0, (w, r, v) -> longVal = v);
+    return longVal;
+  }
+
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:119]: nit: extra empty line

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+  @Override
+  public void skipLongs(int total) {
+    skipValues(total);
+  }
+
+  private void readValues(int total, WritableColumnVector c, int rowId,
+      IntegerOutputWriter outputWriter) {
+    if (valuesRead + total > totalValueCount) {
+      throw new ParquetDecodingException(
+          "no more values to read, total value count is " + valuesRead);
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:212]: nit: let's all use capital letter for the first word to be consistent: "No more values to read ..."

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 19]
```diff
@@ -0,0 +1,315 @@
...
+  @Override
+  public void readLongs(int total, WritableColumnVector c, int rowId) {
+    readValues(total, c, rowId, WritableColumnVector::putLong);
+  }
+
+  @Override
+  public final void readLongsWithRebase(
+      int total, WritableColumnVector c, int rowId, boolean failIfRebase) {
+    readValues(total, c, rowId, (w, r, v) -> {
+      if (v < RebaseDateTime.lastSwitchJulianDay()) {
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:176]: should be `v < RebaseDateTime.lastSwitchJulianTs()`?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    return valuesRead;
+  }
+
+  private void readBlockHeader() {
+    try {
+      minDeltaInCurrentBlock = BytesUtils.readZigZagVarLong(in);
+    } catch (IOException e) {
+      throw new ParquetDecodingException("can not read min delta in current block", e);
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:274]: ditto: "Can not read .."

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+    if (remainingInBlock == 0) {
+      readBlockHeader();
+    }
+
+    // new miniblock, unpack the miniblock
+    if (remainingInMiniBlock == 0) {
+      unpackMiniBlock();
+    }
+
+    //read values from miniblock
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:254]: nit: space after `//`

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 22]
```diff
@@ -0,0 +1,315 @@
...
+    int valuesRead = 0;
+    for (int i = miniBlockSizeInValues - remainingInMiniBlock;
+        i < miniBlockSizeInValues && valuesRead < remaining; i++) {
+      //calculate values from deltas unpacked for current block
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:258]: ditto

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 23]
```diff
@@ -0,0 +1,315 @@
...
+    remainingInMiniBlock = 0;
+  }
+
+  /**
+   * mini block has a size of 8*n, unpack 32 value each time
+   *
+   * see org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesReader#unpackMiniBlock
+   */
+  private void unpackMiniBlock() throws IOException {
+    Arrays.fill(this.unpackedValuesBuffer, 0);
```
Comment (sunchao) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:288]: Is this necessary? since we are overwriting it anyway.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala [hunk 24]
```diff
@@ -16,14 +16,22 @@
...
+import org.apache.hadoop.fs.Path
+import org.apache.parquet.column.{Encoding, ParquetProperties}
 import org.apache.parquet.hadoop.ParquetOutputFormat
 
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.util.DateTimeUtils
+import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSparkSession
 
+
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala:34]: nit: unnecessary change.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala [hunk 25]
```diff
@@ -147,6 +159,16 @@ object DataSourceReadBenchmark extends SqlBasedBenchmark {
           }
         }
 
+        sqlBenchmark.addCase("SQL Parquet Vectorized (Delta Binary)") { _ =>
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:162]: Can we:
- move this up to be adjacent to the Parquet cases
- change the name to "SQL Parquet Vectorized Format v2" or something? 

You also need to rebase and handle the newly added `BooleanType` here.

**Review by sunchao — COMMENTED** [review]
Thanks @parthchandra for the update and sorry for my late response!

The benchmark looks good. I think it's definitely worth doing because otherwise users will get exception instead and have to resort to the much slower `parquet-mr` implementation.

**Review by parthchandra — COMMENTED** [review]
Addressed review comments and rebased

**Review by sunchao — COMMENTED** [review]
Thanks a lot for updating this @parthchandra ! Overall look pretty good. I think we just need to address the issue with benchmark and attach the result together with the PR. You can find out how to get benchmark result using GitHub workflow [here](https://spark.apache.org/developer-tools.html).

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala (hunk 25 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:157]: I guess I didn't understand the implication when you suggested I need to rebase!
It looks like the Boolean type breaks vectorized reading. The boolean column gets encoded as RLE but the vectorized RLE reader does not support Boolean.
```
[info] Caused by: java.lang.UnsupportedOperationException: only readInts is valid.
[info] 	at org.apache.spark.sql.execution.datasources.parquet.VectorizedRleValuesReader.readBooleans(VectorizedRleValuesReader.java:368)
```
I guess I could fix this. Would you recommend a different PR, or should I make the change in this one?

Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:157]: Right, Boolean is encoded as RLE in Parquet format V2, so we'd need to handle it too. I think we should raise a separate PR for this.

For now, can you remove the changes in `DataSourceReadBenchmark` and split it out into a separate PR later once we have the boolean type supported?

**Comment by sunchao** [comment]
@parthchandra could you address the unit tests failure?

**Comment by parthchandra** [comment]
The unit tests are failing in parts that I am not familiar with. Previously, re-running the tests had worked, but this time around the tests are failing every time. Can I get some help figuring out where the problem is?

**Comment by sunchao** [comment]
If you click the SparkQA test build link you should see the failed tests. For instance:

```
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: "[no more values to read, total value count is 641]" did not equal "[No more values to read. Total values read:  641, total count: 641, trying to read 1 more.]"
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetDeltaEncodingSuite.$anonfun$new$21(ParquetDeltaEncodingSuite.scala:170)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
```

**Comment by parthchandra** [comment]
Thank you @sunchao. I had gone thru the log and failed to see the test(s) that had failed. One of the unit tests was checking for the error message that accompanied an exception and as part of the review I had changed the error message! Updated the test. The tests should pass now.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by sunchao** [comment]
Committed to master branch, thanks @parthchandra !

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala [hunk 26]
```diff
@@ -76,6 +79,7 @@ object DataSourceReadBenchmark extends SqlBasedBenchmark {
     saveAsCsvTable(testDf, dir.getCanonicalPath + "/csv")
     saveAsJsonTable(testDf, dir.getCanonicalPath + "/json")
     saveAsParquetTable(testDf, dir.getCanonicalPath + "/parquet")
+    saveAsParquetV2Table(testDf, dir.getCanonicalPath + "/parquetV2")
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:82]: Maybe we should update the benchmark-result of `DataSourceReadBenchmark` @parthchandra parthchandra

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java [hunk 27]
```diff
@@ -0,0 +1,319 @@
...
+  private int currentMiniBlock = 0;
+  private int[] bitWidths; // bit widths for each miniBlock in the current block
+  private int remainingInBlock = 0; // values in current block still to be read
+  private int remainingInMiniBlock = 0; // values in current mini block still to be read
+  private long[] unpackedValuesBuffer;
+
+  private ByteBufferInputStream in;
+
+  // temporary buffers used by readByte, readShort, readInteger, and readLong
+  byte byteVal;
```
Comment (LuciferYang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:76]: Should these 4 field be private?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala [hunk 28]
```diff
@@ -134,33 +134,35 @@ abstract class ParquetRebaseDatetimeSuite
...
+          paths.foreach(_.delete())
         val oldPath = getResourceParquetFilePath("test-data/" + fileName)
         val path3_x = paths(0).getCanonicalPath
         val path3_x_rebase = paths(1).getCanonicalPath
-        val df = Seq.tabulate(N)(rowFunc).toDF("dict", "plain")
-          .select($"dict".cast(catalystType), $"plain".cast(catalystType))
-        withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key -> tsOutputType) {
+          val df = Seq.tabulate(N)(rowFunc).toDF("dict", "plain")
+            .select($"dict".cast(catalystType), $"plain".cast(catalystType))
+          withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key -> tsOutputType) {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:145]: Something wrong with indentation here and below:
Screenshot 2022-01-06 at 22 56 02

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala [hunk 29]
```diff
@@ -219,60 +221,62 @@ abstract class ParquetRebaseDatetimeSuite
...
-                      checkAnswer(
+                  // Force to not rebase to prove the written datetime values are rebased
+                  // and we will get wrong result if we don't rebase while reading.
+                  withSQLConf("spark.test.forceNoRebase" -> "true") {
+                    checkAnswer(
                         spark.read.parquet(path).select($"ts".cast("string")),
                         Seq.tabulate(N)(_ => Row(nonRebased)))
                     }
                   }
+                  }
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:279]: Wrong indentation too

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala (hunk 26 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:82]: I found that there are still unsupported encoding in Data Page V2, such as RLE. It seems that it is not time to update the benchmark, please ignore my previous comments

Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:82]: @LuciferYang I was getting ready to set up a PR for the RLE/Boolean encoding and noticed that you have done so. Thank you!
Adding back the benchmark in a new PR.
